README.md
Dimorphite-DL
=============

This repository is a modified version of [Durrant Lab](https://durrantlab.pitt.edu/)'s original [Dimorphite-DL](https://git.durrantlab.pitt.edu/jdurrant/dimorphite_dl) implementation ([original article](https://www.doi.org/10.1186/s13321-019-0336-9)).

Changes were brought to facilitate its installation and integration in bigger cheminformatics projects (see [CHANGES.md](CHANGES.md)).

What is it?
-----------

As stated on Durrant's Lab original GitLab page: 
<table>
  <tr>
    <td align="justify">
Dimorphite-DL adds hydrogen atoms to molecular representations, as appropriate
for a user-specified pH range. 
It is a fast, accurate, accessible, and modular open-source program for enumerating small-molecule ionization states.  
Users can provide SMILES strings from the command line or via an .smi file.
</td>
  </tr>
</table>

Citation
--------

If you use Dimorphite-DL in your research, please cite:

Ropp PJ, Kaminsky JC, Yablonski S, Durrant JD (2019) Dimorphite-DL: An
open-source program for enumerating the ionization states of drug-like small
molecules. J Cheminform 11:14. doi:[10.1186/s13321-019-0336-9](https://www.doi.org/10.1186/s13321-019-0336-9).

Licensing
---------

Dimorphite-DL is released under the Apache 2.0 license. See LICENCE.txt for
details.

Installation
---------

This modified version of Dimorphite-DL can be installed using [pip](https://pip.pypa.io/en/stable/getting-started/):
```bash
pip install dimorphite-ojmb
```
If the above does not work, you can run the installation like so:
```bash
python -m pip install dimorphite-ojmb
```



Usage
-----

```
usage: dimorphite [-h] [--min_ph MIN] [--max_ph MAX] [--pka_precision PRE]
                  [--smiles SMI] [--smiles_file FILE] [--output_file FILE]
                  [--max_variants MXV] [--label_states] [--silent] [--test]

Dimorphite 1.2.4: Creates models of appropriately protonated small moleucles.
Apache 2.0 License. Copyright 2020 Jacob D. Durrant.

Options:
  -h, --help           Show this message and exit.
  --min_ph MIN         minimum pH to consider (default: 6.4)
  --max_ph MAX         maximum pH to consider (default: 8.4)
  --pka_precision PRE  pKa precision factor (number of standard devations,
                       default: 1.0)
  --smiles SMI         SMILES string to protonate NOTE: This argument is
                       mutually exclusive with smiles_file.  [required]
  --smiles_file FILE   file that contains SMILES strings to protonate NOTE:
                       This argument is mutually exclusive with smiles.
                       [required]
  --output_file FILE   output file to write protonated SMILES (optional)
  --max_variants MXV   limit number of variants per input compound (default:
                       128)
  --label_states       label protonated SMILES with target state (i.e.,
                       "DEPROTONATED", "PROTONATED", or "BOTH").
  --silent             do not print any messages to the screen
  --test               run unit tests (for debugging)
```

The default pH range is 6.4 to 8.4, considered biologically relevant pH.

CLI usage examples
--------

```
  dimorphite --smiles_file sample_molecules.smi
  dimorphite --smiles "CCC(=O)O" --min_ph -3.0 --max_ph -2.0
  dimorphite --smiles "CCCN" --min_ph -3.0 --max_ph -2.0 --output_file output.smi
  dimorphite --smiles_file sample_molecules.smi --pka_precision 2.0 --label_states
  dimorphite --test
```

Advanced Usage
--------------

It is also possible to access Dimorphite-DL from another Python script, rather
than from the command line. Here's an example:

```python
from rdkit import Chem
import dimorphite_dl

# Using the dimorphite_dl.run() function, you can run Dimorphite-DL exactly as
# you would from the command line. Here's an example:
dimorphite_dl.run(smiles="CCCN",
                  min_ph=-3.0,
                  max_ph=-2.0)


# One can also provide multiple SMILES at once.
dimorphite_dl.run(["C[C@](F)(Br)CC(O)=O", "CCCCCN"],
                  min_ph=5.0,
                  max_ph=9.0,
                  silent=True)
```

Caveats
-------

Dimorphite-DL deprotonates indoles and pyrroles around pH 14.5. But these
substructures can also be protonated around pH -3.5. Dimorphite does not
perform the protonation.

Authors and Contacts
--------------------

See the `CONTRIBUTORS.md` file for a full list of contributors. Please contact
Jacob Durrant (durrantj@pitt.edu) with any questions.


CHANGES.md
Changes
=======

1.2.5
-----

* The API has been reworked to allow calling subcomponents of Dimorphite-DL and its packaging.
* Removal of the function `run_with_mol_list` as `run` now supports lists of SMILES.
* The CLI now relies on [Click](https://click.palletsprojects.com/).
* Support input molecules. 


1.2.4
-----

* Dimorphite-DL now better protonates compounds with polyphosphate chains
  (e.g., ATP). See `site_substructures.smarts` for the rationale behind the
  added pKa values.
* Added test cases for ATP and NAD.
* `site_substructures.smarts` now allows comments (lines that start with `#`).
* Fixed a bug that affected how Dimorphite-DL deals with new protonation
    states that yield invalid SMILES strings.
  * Previously, it simply returned the original input SMILES in these rare
    cases (better than nothing). Now, it instead returns the last valid SMILES
    produced, not necessarily the original SMILES.
  * Consider `O=C(O)N1C=CC=C1` at pH 3.5 as an example.
    * Dimorphite-DL first deprotonates the carboxyl group, producing
      `O=C([O-])n1cccc1` (a valid SMILES).
    * It then attempts to protonate the aromatic nitrogen, producing
      `O=C([O-])[n+]1cccc1`, an invalid SMILES.
    * Previously, it would output the original SMILES, `O=C(O)N1C=CC=C1`. Now
      it outputs the last valid SMILES, `O=C([O-])n1cccc1`.
* Improved suport for the `--silent` option.
* Reformatted code per the [*Black* Python code
  formatter](https://github.com/psf/black).

1.2.3
-----

* Updated protonation of nitrogen, oxygen, and sulfur atoms to be compatible
  with the latest version of RDKit, which broke backwards compatibility.
* Added "silent" option to suppress all output.
* Added code to suppress unnecessary RDKit warnings.
* Updated copyright to 2020.

1.2.2
-----

* Added a new parameter to limit the number of variants per compound
  (`--max_variants`). The default is 128.

1.2.1
-----

* Corrected a bug that rarely misprotonated/deprotonated compounds with
  multiple ionization sites (e.g., producing a carbanion).

1.2
---

* Corrected a bug that led Dimorphite-DL to sometimes produce output molecules
  that are non-physical.
* Corrected a bug that gave incorrect protonation states for rare molecules
  (aromatic rings with nitrogens that are protonated when electrically
  neutral, e.g. pyridin-4(1H)-one).
* `run_with_mol_list()` now preserves non-string properties.
* `run_with_mol_list()` throws a warning if it cannot process a molecule,
  rather than terminating the program with an error.

1.1
---

* Dimorphite-DL now distinguishes between indoles/pyrroles and
  Aromatic_nitrogen_protonated.
* It is now possible to call Dimorphite-DL from another Python script, in
  addition to the command line. See the `README.md` file for instructions.

1.0
---

The original version described in:

Ropp PJ, Kaminsky JC, Yablonski S, Durrant JD (2019) Dimorphite-DL: An
open-source program for enumerating the ionization states of drug-like small
molecules. J Cheminform 11:14. doi:10.1186/s13321-019-0336-9.


src/dimorphite_dl/cli.py
# -*- coding: utf-8 -*-

"""This entry point allows the use of dimorphite-dl as a CLI."""

import argparse
from typing import Optional

import click

from .dimorphite_dl import *


examples = """Examples:

    dimorphite --smiles_file sample_molecules.smi
    dimorphite --smiles "CCC(=O)O" --min_ph -3.0 --max_ph -2.0
    dimorphite --smiles "CCCN" --min_ph -3.0 --max_ph -2.0 --output_file output.smi
    dimorphite --smiles_file sample_molecules.smi --pka_precision 2.0 --label_states
    dimorphite --test"""


class Mutex(click.Option):
    def __init__(self, *args, **kwargs):
        """Custom class allowing click.Options to be
        required if other click.Options are not set.

        Derived from: https://stackoverflow.com/a/61684480
        """
        self.not_required_if: list = kwargs.pop("not_required_if")

        assert self.not_required_if, "'not_required_if' parameter required"
        assert isinstance(self.not_required_if, list), "'not_required_if' mut be a list"
        kwargs["help"] = (kwargs.get("help", "") + ' NOTE: This argument is mutually exclusive with ' + ", ".join(
            self.not_required_if) + ".").strip()
        super(Mutex, self).__init__(*args, **kwargs)

    def handle_parse_result(self, ctx, opts, args):
        """Override base method."""
        current_opt: bool = self.consume_value(ctx, opts)[0]  # Obtain set value
        for other_param in ctx.command.get_params(ctx):
            if other_param is self:  # Ignore the current parameter
                continue
            # Other argument's name or declaration in self.not_required_if
            if other_param.human_readable_name in self.not_required_if or any(
                    opt.lstrip('-') in self.not_required_if for opt in other_param.opts) or any(
                    opt.lstrip('-') in self.not_required_if for opt in other_param.secondary_opts):
                # Get value assigned to the other option
                other_opt: bool = other_param.consume_value(ctx, opts)[0]
                if other_opt:
                    if current_opt:
                        raise click.UsageError(
                            "Illegal usage: '" + str(self.name)
                            + "' is mutually exclusive with '"
                            + str(other_param.human_readable_name) + "'."
                        )
                    else:
                        self.required = None  # Override requirement
        return super(Mutex, self).handle_parse_result(ctx, opts, args)


class SpecialEpilog(click.Group):
    """Custom class with a custom display of the epilog."""
    def format_epilog(self, ctx, formatter):
        if self.epilog:
            formatter.width = float('inf')
            formatter.write_paragraph()
            for line in self.epilog.split('\n'):
                formatter.write_text(line)

    def get_help(self, ctx):
        """ standard get help, but without rstrip """
        formatter = ctx.make_formatter()
        self.format_help(ctx, formatter)
        return formatter.getvalue()

    def format_usage(self, ctx, formatter):
        pieces = ["[-h]", "[--min_ph MIN]", "[--max_ph MAX]", "[--pka_precision PRE]", "[--smiles SMI]",
                  "[--smiles_file FILE]", "[--output_file FILE]", "[--max_variants MXV]", "[--label_states]",
                  "[--silent]", "[--test]"]
        formatter.write_usage(ctx.command_path, " ".join(pieces))
        formatter.write_text(" ")
        formatter.write_text(HEADER)
        formatter.write_text(" \n")


@click.command(help="Dimorphite 1.2.4: Creates models of appropriately protonated small molecules.\n"
                    "Apache 2.0 License. Copyright 2020 Jacob D. Durrant.",
               epilog=examples,
               cls=SpecialEpilog, invoke_without_command=True,
               context_settings={'help_option_names': ['-h', '--help']})
@click.option("--min_ph", metavar="MIN", type=float, default=6.4, required=False,
              help="minimum pH to consider (default: 6.4)")
@click.option("--max_ph", metavar="MAX", type=float, default=8.4, required=False,
              help="maximum pH to consider (default: 8.4)")
@click.option("--pka_precision", metavar="PRE", type=float, default=1.0, required=False,
              help="pKa precision factor (number of standard devations, default: 1.0)")
@click.option("--smiles", metavar="SMI", type=str, required=True, cls=Mutex, not_required_if=['smiles_file', 'test'],
              help="SMILES string to protonate", multiple=True)
@click.option("--smiles_file", metavar="FILE", type=str, required=True, cls=Mutex, not_required_if=['smiles', 'test'],
              help="file that contains SMILES strings to protonate")
@click.option("--output_file", metavar="FILE", type=str, required=False,
              help="output file to write protonated SMILES (optional)")
@click.option("--max_variants", metavar="MXV", type=int, default=128, required=False,
              help="limit number of variants per input compound (default: 128)")
@click.option("--label_states", is_flag=True, required=False, default=False,
              help="label protonated SMILES with target state (i.e., \"DEPROTONATED\", \"PROTONATED\", or \"BOTH\").")
@click.option("--silent", is_flag=True, required=False, default=False,
              help="do not print any messages to the screen")
@click.option("--test", is_flag=True, required=False, default=False,
              help="run unit tests (for debugging)")
def main(min_ph: float = 6.4,
         max_ph: float = 8.4,
         pka_precision: float = 1.0,
         smiles: Optional[str] = None,
         smiles_file: Optional[str] = None,
         output_file: Optional[str] = None,
         max_variants: int = 128,
         label_states: bool = False,
         silent: bool = False,
         test: bool = False):
    if test:
        TestFuncs.test()
        return
    # Main part
    smiles = list(smiles)
    if len(smiles) > 0:
        smiles_file = StringIO('\n'.join(smiles))
    # Read in data in SMI format
    smiles_and_data = LoadSMIFile(smiles_file, silent=silent)
    # Separate SMILES from data
    all_smis, all_data = zip(*[[smi_and_data['smiles'], smi_and_data['data']] for smi_and_data in smiles_and_data])
    all_smis, all_data = list(all_smis), list(all_data)
    # Run protonation
    protonated_smiles = run(all_smis, min_ph=min_ph, max_ph=max_ph, pka_precision=pka_precision,
                            max_variants=max_variants, label_states=label_states,
                            silent=silent)
    if label_states:
        for i in range(len(protonated_smiles)):
            if protonated_smiles[i] is not None:
                label = protonated_smiles[i][1]
                data = all_data[i]
                for smi in protonated_smiles[i][0]:
                    line = f"{smi}"
                    if len(data) > 0:
                        line += '\t' + '\t'.join(data)
                    if label is not None:
                        if isinstance(label, list):
                            line += '\t' + '\t'.join(label)
                        else:
                            line += f"\t{label}"
                    line += '\n'
                    if output_file is not None:
                        # An output file was specified, so write to that.
                        with open(output_file, "a") as oh:
                            print(line, file=oh)
                    else:
                        print(line, end='')
    else:
        for i in range(len(protonated_smiles)):
            if protonated_smiles[i] is not None:
                data = all_data[i]
                line = [protonated_smiles[i]] if isinstance(protonated_smiles[i], str) \
                    else protonated_smiles[i]
                if len(data) > 0:
                    line = '\n'.join([f"{smi}\t" + '\t'.join(data) for smi in line])
                else:
                    line = '\n'.join(line)
                line += '\n'
                if output_file is not None:
                    # An output file was specified, so write to that.
                    with open(output_file, "a") as oh:
                        print(line, file=oh)
                else:
                    print(line, end='')
